### PR TITLE
Game State - simplified solution

### PR DESCRIPTION
**Pull request by Eskimo396**
Overhauled the game state loop in main.py to simplify the update and drawing workflow of the game. This is one suggested way of re-working the game state loop. I found the original loop in main.py to be overly complicated and limited in the potential for expansion. This solution is the simplest version I could think of and makes it clear where to include new drawing and update features in the future. Having less duplicated code will benefit us in the long run by making it easier to locate functions and make changes. 

-simplified from 2 classes in main.py to 1 game class
-removed multiple duplicate initializations and variables
-set up main menu, pause menu and settings menu in match/case branches in the main game loop.
-split update functions from several classes into separate draw and update functions
-removed duplicate music being started in level.py
-combined setup functions into init on several classes as they are only ever called once and initiate variables for the entire class

-1 known error - slider in settings menu - causes large amount of lag when going through the menu. Has been commented out until a fix can be found

### CODE REVIEW

**Comment by Eskimo396** [comment]
I'm not sure how to fix the lint failure. Looks as though it doesn't like the import * and is therefore throwing errors?

**Comment by ultimateownsz** [comment]
Match case is really powerful but, in scope of complexity can it be difficult to understand for people who have less experience with coding. Because in the project you don't need prior experience, so it adds a learning curve to it.

For the complexity in the game with all the different states, match cases can help to some degree but just wanted to give some road blocks newer programmers can encounter.

If we use more complexity than simplicity in the code overall.

**Comment by fortwoone** [comment]
…I miiiiiiiiight have used match-case somewhere to replace an if chain

**Comment by Eskimo396** [comment]
I didn't think of match case being too complex. It's a base feature of most programming languages (typically called switch case). It is usually included in beginner guides when learning syntax. It seemed a lot simpler as the readability is so much better than numerous cascading if statements. If we choose to keep this implementation, I think adding in some comments to explain what is happening will resolve any experience issues that arise.

File: main.py [hunk 1]
```diff
@@ -1,142 +1,129 @@
...
+        self.screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
         pygame.init()
-        self.screen = pygame.display.set_mode((
-            settings.SCREEN_WIDTH,
-            settings.SCREEN_HEIGHT,
-        ))
         pygame.display.set_caption('PyDew')
+
+        # variables 
+        self.gamestate = "main menu"
```
Comment (fortwoone) [inline main.py:19]: I suggest using IntEnums instead of strings, because it keeps the code consistent. And you will encounter no typo problem if you use enum constants instead of strings

File: main.py [hunk 2]
```diff
@@ -1,142 +1,129 @@
...
-                    pygame.quit()
-                    sys.exit()
             keys = pygame.key.get_just_pressed()
+
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:     
+                    self.quit()
+
+            # UPDATE -----------------------------------
+            match self.gamestate:
```
Comment (fortwoone) [inline main.py:59]: See my previous comment about enums.

File: src/groups.py [hunk 3]
```diff
@@ -1,11 +1,5 @@
 import pygame
-from src.settings import (
-    LAYERS,
-    SCREEN_HEIGHT,
-    SCREEN_WIDTH,
-    Coordinate,
-)
-
+from src.settings import *
```
Comment (fortwoone) [inline src/groups.py:2]: Wildcard imports will make the linter complain, so it's best that this gets reverted

File: src/level.py [hunk 4]
```diff
@@ -7,19 +7,8 @@
...
-    Sprite,
-    Player,
-)
-from src.settings import (
-    TILE_SIZE,
-    SCALE_FACTOR,
-    LAYERS,
-    MapDict,
-)
+from src.sprites import *
```
Comment (fortwoone) [inline src/level.py:10]: See my previous comment about wildcard imports

File: src/level.py [hunk 5]
```diff
@@ -187,7 +189,7 @@ def reset(self):
         # plants
         self.soil_layer.update_plants()
 
-        self.sky.set_time(6, 0)  # set to 0600 hours upon sleeping
+        self.sky.set_time(0, 0)  # set to 0600 hours upon sleeping
```
Comment (fortwoone) [inline src/level.py:192]: ...why did you change that? The game is going to boot back at night now...

File: src/main_menu.py [hunk 6]
```diff
@@ -1,18 +1,15 @@
 import pygame
-from src.settings import (
-    SCREEN_HEIGHT,
-    SCREEN_WIDTH,
-)
+from src.settings import *
```
Comment (fortwoone) [inline src/main_menu.py:2]: See my comment about wildcard imports

File: src/menu.py [hunk 7]
```diff
@@ -1,10 +1,5 @@
 import pygame
-from src.settings import (
-    SCREEN_HEIGHT,
-    SCREEN_WIDTH,
-    PURCHASE_PRICES,
-    SALE_PRICES,
-)
+from src.settings import *
```
Comment (fortwoone) [inline src/menu.py:2]: Insert my message about wildcard imports here

File: src/overlay.py [hunk 8]
```diff
@@ -1,7 +1,6 @@
 import pygame
-from src.settings import (
-    OVERLAY_POSITIONS,
-)
+from src.settings import *
```
Comment (fortwoone) [inline src/overlay.py:2]: ...and again, wildcard imports

File: src/pause_menu.py [hunk 9]
```diff
@@ -1,13 +1,9 @@
 import pygame
-from src.settings import (
-    SCREEN_HEIGHT,
-    SCREEN_WIDTH,
-)
+from src.settings import *
```
Comment (fortwoone) [inline src/pause_menu.py:2]: Wildcaaaaards

File: src/settings_menu.py [hunk 10]
```diff
@@ -1,8 +1,5 @@
 import pygame
-from src.settings import (
-    SCREEN_HEIGHT,
-    SCREEN_WIDTH,
-)
+from src.settings import *
```
Comment (fortwoone) [inline src/settings_menu.py:2]: ....yep, you guessed it, insert my wildcard import comment again

File: src/soil.py [hunk 11]
```diff
@@ -1,14 +1,6 @@
...
-    LAYERS,
-)
-from src.sprites import (
-    Plant,
-    Sprite,
-)
-import random
+from .settings import *
+from .sprites import Sprite, Plant
+from random import *
```
Comment (fortwoone) [inline src/soil.py:3]: And again, wildcard imports!

File: src/sprites.py [hunk 12]
```diff
@@ -1,19 +1,10 @@
-import pygame
+from .settings import *
```
Comment (fortwoone) [inline src/sprites.py:1]: Insert my wilcard import comment again

File: src/support.py [hunk 13]
```diff
@@ -1,137 +1,106 @@
...
-import pytmx
-import sys
-from src import settings
-from src.settings import (
-    CHAR_TILE_SIZE,
-    SCALE_FACTOR,
-    TILE_SIZE,
-)
+from .settings import *
+from os import *
```
Comment (fortwoone) [inline src/support.py:2]: And again!

File: src/timer.py [hunk 14]
```diff
@@ -1,5 +1,5 @@
 import pygame
-
+from .settings import *
```
Comment (fortwoone) [inline src/timer.py:2]: What exactly do you use from settings here? No need to import anything from there, is there?

**Review by fortwoone — CHANGES_REQUESTED** [review]
The gist of the solution looks good to me, but the execution needs to changed, see the multiple comments I placed on the code

**Comment by fortwoone** [comment]
> I'm not sure how to fix the lint failure. Looks as though it doesn't like the import * and is therefore throwing errors?

Basically, yes. Linters follow the PEP 8 writing style convention, which explicitly deems wildcard imports (with the *) bad. The reason for that is that while you don't need to add giant lists of imports if you use virtually all names in a file, if you have one name referring to two different things in two files, that can not only make it unclear for the interpreter itself when the game runs, but also for YOURSELF. Maybe you understand where a name comes from right now, but if you stop coding and come back later on, for example, you could very well have forgotten that. Hence, it's better to use either normal imports or use from-based imports with name lists

File: src/level.py (hunk 5 above)
Comment (Eskimo396) [reply src/level.py:192]: Ah this was a test just to make sure I didn't screw up something and the darkening of the day was turned off. I forgot to change it back to 6.

File: main.py (hunk 1 above)
Comment (Eskimo396) [reply main.py:19]: I don't quite understand the benefit of using enums instead of strings. Sure you have to make sure you spell the variable correctly with strings but you already have to do that with any variable name. Enums on the other hand require another library to be imported, have to define every option to a number, and then require a call to the variable like variablename.value. strings definitions are a single line and infinitely expandable without requiring any additional libraries or multiple lines of numbering. I've been programming for years and this is the first time I've come across enums. I had to read up on them and still don't quite understand the use. if we are wanting to keep the code simple for beginners, I'm not sure enums are the simplest way.

File: src/groups.py (hunk 3 above)
Comment (Eskimo396) [reply src/groups.py:2]: I will go back through and import only the required imports.

File: src/timer.py (hunk 14 above)
Comment (Eskimo396) [reply src/timer.py:2]: After I created this PR, when I synced my code, all the headers changed and broke a lot of the code. I went back through and just added the wildcard headers everywhere they had been before to fix the errors. I may have accidentally imported them here thinking it was required.

File: src/level.py (hunk 5 above)
Comment (fortwoone) [reply src/level.py:192]: Okay then, I'll trust you to revert that change

File: main.py (hunk 1 above)
Comment (fortwoone) [reply main.py:19]: Enums are constant names, so if you use completion helpers like in PyCharm for example you can easily select a constant between all available ones for an enum. Also, strings are typo-prone, so if for example your code set `self.gamestate` to `"main myenu"` somehow due to a typo, the game would break, since nothing would be planned for a value of `"main myenu"` as a separate game state. And it would be extremely painful to check where the typo occurred, as you might not even get an error for that. If you use enums and end up making a typo, the interpreter will actively signal it to you by raising a `NameError`, and the game will stop. Moreover, the resulting traceback will show you the exact line where the typo was
